service/history/historybuilder/event_store.go

```diff
@@ -419,7 +419,14 @@ func (b *EventStore) reorderBuffer(
 			enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_FAILED,
 			enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_TIMED_OUT,
 			enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_CANCELED,
-			enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_TERMINATED:
+			enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_TERMINATED,
+			// TODO: This implementation detail is hurting extensibility of workflows and the ability to add external
+			// components.
```

Comment:
Why do we need to do this?

Reply:
Looks like we want to get rid of this but there's speculation that Cassandra reorders buffered events and we want to make sure that certain events are reordered back in a predictable way.

Essentially, signals are expected to come before other event types. Not sure if there are other events that get buffered that need this special treatment.